### DISCUSSION TITLE
The city, while expanding, blocks waterways, which may block the movement of ships

### DISCUSSION THREAD

**Discussion by LC-Zorg** [OpenTTD/OpenTTD#11735] [Features]
## Version of OpenTTD
1.11.2 (from always)

## Actual result
The city can build on tiles that are half land and half water, which in the case of even quite wide canals running near the city, that as they develop, they build over these canals over time, closing the water transport route.
![zzz 3](https://user-images.githubusercontent.com/67662046/120104067-037e7e80-c153-11eb-94b4-9e74822d0053.png)

## Expected result
The city cannot build on tiles that are half land and half water - the city doesn't block the movement of ships.
![ZZZZ](https://user-images.githubusercontent.com/67662046/120104217-a20adf80-c153-11eb-86c0-0ffe808bc7af.png)

## Comments
In the case of buildings and objects, a separate type could be added that would take up half a field - such buildings and objects the city could still build.
In the case of bridges, a different solution to the problem is possible - the city will still be able to build bridges, but these will not occupy a water field - this is a separate issue.

**Comment by 2TallTyler** [comment]
Related: #6935 and #6931.

**Comment by LC-Zorg** [comment]
I looked at the related PR and have some thought of, hopefully, simple solution.

First, let me quote some comments from the PR #6931 created by SamuXarick.

@nielsmh comented [on 6 Jan 2019](https://github.com/OpenTTD/OpenTTD/pull/6931#issuecomment-451754441)
> ![image](https://user-images.githubusercontent.com/67662046/173202655-4cae5668-00fb-4c82-9df9-64d60a0eea44.png)
> Which of these half water tiles would be appropriate to prevent cities from building on, and why?


@Eddi-z commented [on 8 Jan 2019](https://github.com/OpenTTD/OpenTTD/pull/6931#issuecomment-452272595)
> i think your approach to that is overly complicated.
> 
> all this needs is to look at max. 3 other tiles:
> 
> * each water halftile has two adjacent tiles on water level
> * if one of those tiles is not a traversable water tile, allow overbuilding
> * if both are traversable water tiles, look at the other tile that is adjacent to both (diagonal corner of the original tile)
> * if that is also a traversable tile, allow overbuildng
> * if not, deny overbuilding

@LordAro commented [on 20 Jan 2019](https://github.com/OpenTTD/OpenTTD/pull/6931#issuecomment-455899039)
> https://user-images.githubusercontent.com/43006711/50809762-aec9d700-12fd-11e9-8ea9-5e55d3eb8a15.png
> I'm unsure about this as well, but it would be very nice to be able to stop towns from overbuilding their own rivers, like in the image above

@TrueBrain commented [on 3 Mar 2019](https://github.com/OpenTTD/OpenTTD/pull/6931#issuecomment-469012054)
> The expectation is, that this can also be resolved in a few lines of readable code, instead of a huge function which does tons of things. It might to exactly what it should do, but it is very hard to verify, and it is like bringing a gun to a pillow-fight.

I don't know if it is simple to do, although theoretically it seems to be like that.
The idea is an even more simplified version of what Eddi described. It consists of checking only one tile in front of the one on which the city would put the building or road. The city may only building on this tile if there is a full, clear water area on the specific adjacent tile.
![One tile solution](https://user-images.githubusercontent.com/67662046/173203039-3a8fbe57-715e-41c1-8b20-864f15c9995a.png)

For the case of bridges, this could be a separate solution.
Here I see three options:
1. The city cannot build bridges on half-water tiles - the simplest, though perhaps too restrictive solution.
2. Before building the bridge (if it is possible to build there), the city also checks the tile at the other end. If construction on it is possible, the city builds a bridge, if not, a road section - I don't know if the simplest, but it seems to be the surest solution.
3. Before building the bridge (if it is possible to build it in this place), appropriate tiles are checked in specific locations, which gives the answer whether the bridge is not built along the quay or too close to the shore - this solution will not eliminate all cases, but it would also reduce another problem as is building bridges in places that don't make sense.
https://user-images.githubusercontent.com/67662046/170138711-0e89de24-a708-4a35-8bba-240d3a2ed90d.png

A certain drawback of main solution may be the cases of fields marked in orange. Theoretically, there are no reasons why the city cannot build in such cases. But solving this problem would require not only checking three fields, but also taking into account many different possibilities and layouts, which in my opinion would be unprofitable. Moreover, a situation where a part of the quay is free may be beneficial for the player and allow him to build a port. However, when he thinks that it is unsightly, he can level the area and the city will build up there.
